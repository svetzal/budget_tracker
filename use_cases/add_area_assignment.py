from datetime import date
from typing import Optional

from entities import AreaAssignment, LEVEL_PROGRAM, LEVEL_EXECUTIVE
from use_cases.use_case import UseCase


class AddAreaAssignment(UseCase):
    def execute(self, support_area_code: str, person_code: str, funding_source_transit: int, start_date: str,
                end_date: str, level: Optional[str]):
        self.guard_support_area_must_exist(support_area_code)
        self.guard_person_must_exist(person_code)
        self.guard_funding_source_must_exist(funding_source_transit)
        if level not in [LEVEL_EXECUTIVE, LEVEL_PROGRAM]:
            raise ValueError(f"Level must be 'exec' or 'program', not {level}")
        area_assignment = AreaAssignment(
            support_area_code=support_area_code,
            person_code=person_code,
            funding_source_transit=funding_source_transit,
            level=level,
            start_date=date.fromisoformat(start_date),
            end_date=date.fromisoformat(end_date)
        )
        self.practice.area_assignments.append(area_assignment)
