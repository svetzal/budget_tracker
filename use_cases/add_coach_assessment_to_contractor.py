from entities import CoachAssessment
from use_cases.use_case import UseCase


class AddCoachAssessmentToContractor(UseCase):
    def execute(self, contractor_code: str, completed_on: str, leadership: float, technical: float, practice: float):
        coach = self.find_contractor(contractor_code)
        coach.assessments.append(
            CoachAssessment(
                completed_on=completed_on,
                leadership=leadership,
                technical=technical,
                practice=practice,
            )
        )
