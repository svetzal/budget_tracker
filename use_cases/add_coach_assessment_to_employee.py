from entities import CoachAssessment
from use_cases.use_case import UseCase


class AddCoachAssessmentToEmployee(UseCase):
    def execute(self, employee_code: str, completed_on: str, leadership: float, technical: float, practice: float):
        coach = self.find_employee(employee_code)
        coach.assessments.append(
            CoachAssessment(
                completed_on=completed_on,
                leadership=leadership,
                technical=technical,
                practice=practice,
            )
        )
