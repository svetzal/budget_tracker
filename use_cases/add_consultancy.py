from entities import Consultancy
from use_cases.use_case import UseCase


class AddConsultancy(UseCase):
    def execute(self, code: str, name: str, contract: str, contact_name: str, contact_phone: str, contact_email: str):
        self.guard_consultancy_duplicate(code)
        consultancy = Consultancy(
            code=code,
            name=name,
            contract=contract,
            contact_name=contact_name,
            contact_phone=contact_phone,
            contact_email=contact_email
        )
        self.practice.consultancies.append(consultancy)
