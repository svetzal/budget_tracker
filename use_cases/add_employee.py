import datetime

from entities import Employee
from use_cases.use_case import UseCase


class AddEmployee(UseCase):
    def execute(self, code: str, name: str, email: str, start_date: str, end_date: str = None, phone_number: str = None):
        self.guard_employee_not_duplicate(code)
        employee = Employee(
            code=code,
            name=name,
            email=email,
            start_date=start_date,
            phone_number=phone_number,
            end_date=datetime.date.fromisoformat(end_date) if end_date else None,
        )
        self.practice.employees.append(employee)
