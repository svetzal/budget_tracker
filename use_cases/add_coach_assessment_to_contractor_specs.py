from entities import CoachingPracticeFinance
from use_cases.add_coach_assessment_to_contractor import AddCoachAssessmentToContractor
from use_cases.test_data_generators import FIRST_CONTRACTOR_CODE
from verifiers import verify_model


def test_can_add_coach_assessment_to_contractor(staffed_practice: CoachingPracticeFinance):
    AddCoachAssessmentToContractor(staffed_practice).execute(
        contractor_code=FIRST_CONTRACTOR_CODE,
        completed_on="2023-01-01",
        leadership=0.5,
        technical=0.5,
        practice=1.0,
    )

    verify_model(staffed_practice.contractors)
