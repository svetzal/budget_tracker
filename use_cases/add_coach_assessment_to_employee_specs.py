from entities import CoachingPracticeFinance
from use_cases.add_coach_assessment_to_employee import AddCoachAssessmentToEmployee
from use_cases.test_data_generators import FIRST_EMPLOYEE_CODE
from verifiers import verify_model


def test_can_add_coach_assessment_to_employee(staffed_practice: CoachingPracticeFinance):
    AddCoachAssessmentToEmployee(staffed_practice).execute(
        employee_code=FIRST_EMPLOYEE_CODE,
        completed_on="2023-01-01",
        leadership=0.5,
        technical=0.5,
        practice=1.0,
    )

    verify_model(staffed_practice.employees)
