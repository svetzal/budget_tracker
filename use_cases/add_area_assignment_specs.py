from _decimal import Decimal

from entities import CoachingPracticeFinance
from model_verifier import verify_model
from use_cases.add_area_assignment import AddAreaAssignment
from use_cases.add_consultancy import AddConsultancy
from use_cases.add_contractor import AddContractor
from use_cases.add_funding_source import AddFundingSource
from use_cases.add_support_area import AddSupportArea


def test_add_area_assignment():
    practice = CoachingPracticeFinance()

    AddSupportArea(practice).execute(
        code="1",
        name="Area 1"
    )

    AddFundingSource(practice).execute(
        transit=1,
        name="Funding Source 1",
        total=Decimal(100000),
        start_date="2020-01-01",
        end_date="2020-12-31",
    )

    AddConsultancy(practice).execute(
        code="1",
        name="Consultancy 1",
        contract="CTR00123",
        contact_name="Someone",
        contact_phone="647-555-1212",
        contact_email="someone@somewhere.com",
    )

    AddContractor(practice).execute(
        code="1",
        name="Contractor 1",
        consultancy_code="1",
        phone_number="647-555-1212",
        email="someone@somewhere.com",
        start_date="2020-01-01",
    )

    AddAreaAssignment(practice).execute(
        support_area_code="1",
        contractor_code="1",
        funding_source_transit=1,
        start_date="2020-01-01",
        end_date="2020-12-31",
    )

    verify_model(practice.area_assignments)