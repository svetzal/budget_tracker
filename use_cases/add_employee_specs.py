from entities import CoachingPracticeFinance
from verifiers import verify_model
from use_cases.add_employee import AddEmployee


def test_can_add_employee(empty_practice, first_employee_details: dict):
    practice = CoachingPracticeFinance()

    AddEmployee(practice).execute(**first_employee_details)

    verify_model(practice.employees)
