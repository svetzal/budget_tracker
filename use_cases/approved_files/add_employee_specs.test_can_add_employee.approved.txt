[
    {
        "code": "1",
        "name": "Employee 1",
        "email": "employee1.name@mycompany.com",
        "start_date": "2021-01-01",
        "phone_number": null,
        "end_date": null
    }
]
